docs/backend-system/core-services/01-index.md

```diff
@@ -65,10 +88,366 @@ createBackendPlugin({
+
+```ts
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+
+createBackendPlugin({
+  id: 'example',
+  register(env) {
+    env.registerInit({
+      deps: {
+        log: coreServices.logger,
+      },
+      async init({ log }) {
+        log.warn('Heres a nice log line thats a warning!');
+      },
+    });
+  },
+});
+```
+
+## Cache
+
+There's a core service provided with the backend system that can be used to interact with a cache in your plugins. This cache is bound to your plugin too, so that you will only set and get values in your plugins namespace.
+
+### Using the service
+
+The following example shows how to get a cache client in your `example` backend plugin and `set` and `get` values from the cache.
+
+```ts
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+
+createBackendPlugin({
+  id: 'example',
+  register(env) {
+    env.registerInit({
+      deps: {
+        cache: coreServices.cache,
+      },
+      async init({ cache }) {
+        const { key, value } = { key: 'test:key', value: 'bob' };
+        await cache.set(key, value, { ttl: 1000 });
+
+        // .. some other stuff.
+
+        await cache.get(key); // 'bob'
+      },
+    });
+  },
+});
+```
+
+## Database
+
+Interacting with databases inside your plugin is something that is quite common, and we provide a `PluginDatabaseManager` part of the core services that you can get `knex` client hooked up to your database which is configured in `app-config.yaml`.
+
+If there's no config provided in `backend.database` then you will automatically get a simple in memory `sqlite3` client for your plugin.
+
+These `PluginDatabaseManager`s are scoped per plugin too, so that table names do not conflict across plugins either.
+
+### Using the service
+
+The following example shows how to get a `PluginDatabaseManager` in your `example` backend plugin and get a `client` for interacting with the database and running some migrations from a `migrationsDir` for your plugin.
+
+```ts
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+
+createBackendPlugin({
+  id: 'example',
+  register(env) {
+    env.registerInit({
+      deps: {
+        database: coreServices.database,
+      },
+      async init({ database }) {
+        const client = database.getClient();
+
+        if (!database.migrations?.skip) {
+          await client.migrate.latest({
+            directory: migrationsDir,
+          });
+        }
+      },
+    });
+  },
+});
+```
+
+## Discovery
+
+When building plugins, you might find that you will need to lookup where in fact another plugins `baseUrl`. This could be for example, a `http` route or some `ws` protocol URL. For this we have the `discovery` service that you can query both the internal and external `baseUrl`s given a plugin ID.
```

Comment:
```suggestion
When building plugins, you might find that you will need to look up another plugin's base URL to be able to communicate with it. This could be for example an HTTP route or some `ws` protocol URL. For this we have a discovery service which can provide both internal and external base URLs for a given a plugin ID.
```